Harmony/Classes/Repository/DebugDataSource.swift

```diff
@@ -0,0 +1,152 @@
+//
+// Copyright 2019 Mobile Jazz SL
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+
+import Foundation
+
+public enum DebugDataSourceDelay {
+    case none
+    case sync(TimeInterval)
+    case async(TimeInterval, DispatchQueue)
+}
+
+public enum DebugDataSourceError {
+    case none
+    case error(Error, probability: Double)
+    
+    func fail() throws {
+        switch self {
+        case .none:
+            break
+        case .error(let error, let probability):
+            if Double.random(in: 0.0...1.0) <= probability {
+                throw error
+            }
+        }
+    }
+}
+
+private class DebugDataSourceToken {
+    let id = String(randomOfLength: 8).uppercased()
+    private var startTime : DispatchTime?
+    private var endTime : DispatchTime?
+    
+    func start() {
+        startTime = DispatchTime.now()
+    }
+    
+    func end() {
+        endTime = DispatchTime.now()
+    }
+    
+    func time() -> TimeInterval? {
+        guard let start = startTime, let end = endTime else {
+            return nil
+        }
+        let seconds = TimeInterval((end.uptimeNanoseconds - start.uptimeNanoseconds)) / 1_000_000_000
+        return seconds
+    }
+}
+
+public class DebugDataSource <D,T> : GetDataSource, PutDataSource, DeleteDataSource where D: GetDataSource, D:PutDataSource, D:DeleteDataSource, D.T == T {
+    
+    private let dataSource : D
+    private let delay : DebugDataSourceDelay
+    private let error : DebugDataSourceError
+    private let logger : Logger
+    
+    public init(_ dataSource: D,
+                delay: DebugDataSourceDelay = .none,
+                error: DebugDataSourceError = .none,
+                logger: Logger = DeviceConsoleLogger()) {
+        self.dataSource = dataSource
+        self.delay = delay
+        self.error = error
+        self.logger = logger
+    }
+    
+    private func postprocess<K>(_ future: Future<K>, _ method: DataSourceCRUD, _ token: DebugDataSourceToken) -> Future<K> {
+        return addError(addDelay(future.onCompletion { token.end() })).then { result in
+            self.log(method, token, "Completed with result: \(String(describing: result))")
+            }.fail{ error in
+                self.log(method, token, "Failed with error: \(error)")
+        }
+    }
+    
+    private func addDelay<K>(_ future: Future<K>) -> Future<K> {
```

Comment:
Which are the benefits of having a delay sync ore async? 

We are always handling the threading in the interactor layer, this behaviour it could lead us to threading issues if it's not handle properly.

Reply:
Futures have a way to delay a response. The API gives the option to do a sync delay (blocking the thread) or async delay (where a `queue` must be provided to continue the execution).

I honestly don't care very much in issues because of this. Programers must use it correctly, and if the UI is wrongly dipsatching in a wrong thread, it is the programer's fault. 

I don't see the existence of these methods as an issue or a problem at all.